74906 hairpin text

MarcSabatella opened the pull request:
Two usability enhancements for the new textline hairpins:

1) prepopulate the begin text for ordinary (non-textline) hairpins (even though it won't be displayed by default) so it is all ready to go if you change it into a textline hairpin via the Inspector

2) adjust vertical position of the textline hairpin by 0.5sp to better align visually with the corresponding ordinary hairpin

The second of these is a bit of a hack, but one designed to correct for something done for 2.0 that was itself a bit of a hack.  The current default position of the ordinary hairpin (a general style setting) was chosen to make the vertical center of the hairpin align with a the vertical center of a dynamic marking at the default position (a text style setting).  These two defaults happen to differ by 0.5sp, because dynamics align to baseline but ordinary hairpins align to the vertical center (the tip) and moving the latter up by 0.5sp made it hit the vertical center of the dynamic with the default fonts.

We *could* add a new style parameter for the vertical position for the textline hairpin (also one to set the line style) and use this instead of the ordinary hairpin default position.  We could also revisit the default style settings and how we interpret the hairpin vertical position - making it specify the baseline position rather than the tip position, and adjust older scores accordingly.

MarcSabatella commented:
I had another thought regarding the vertical position.  If the goal is to align with dynamics, why not simply use the vertical offset from the dynamics text style?  So I updated the second commit to do that.  Seems less hacky, and simpler for the user than adding another style setting and then having to coordinate the dynamics text style and the textline hairpin style.

MarcSabatella commented:
Seems more appropriate to me to match the size of dynamics.  Looking at examples in the literature, that seems how it is most often done, although to be sure, I see text sizes all over the map.

ZackTheCardshark commented:
Seems more appropriate to me to be smaller than dynamics, because dynamics are deliberately very large. It doesn't seem right for this to take up that much space—especially when the text is changed to (for example) "cresc. poco a poco".

MarcSabatella commented:
There is some logic to that, but again, when I checked actual published scores, it seemed more common to have them the same size.  Anyhow, feel free to start a discussion in forum to see if a different consensus emerges.

lasconic commented:
Looks good. Gould seems to agree that cresc. and dynamics are the same size.

Would it be a good idea to set the continue text for *cresc* and *dim* lines? to *(cresc.)* and *(dim.)*

MarcSabatella commented:
Updated to add the continue text, also fixed a glitch where score text style was not being honored for hairpins added via double click (I just needed to add hairpins to the list of spanners covered by the style rebase code in cmdAddSpanner).  So now if you set your hairpin text style to 10pt, you'll get what you want.  Also I squashed the commits.

MarcSabatella commented:
@ZackTheCardshark - see #2202.  The tests would need to be updated for this to be merged.  Seems like a great learning opportunity for someone who wants to see this happen... :-)

ZackTheCardshark commented:
See… what? I don't even know what I'm looking at. Do you mean https://travis-ci.org/musescore/MuseScore/builds/78607070 ?

MarcSabatella commented:
Yes.  It shows a bunch of tests that failed becausethey contained hairpins that would now contain text whereas they didn't before, so the tests need to be updated to expect that text.  It's mostly a process of going through the specified folders and copying the generated mscx file to the reference file for each of the failures.  It's not really fun, but like I said, a learning opportunity!

Not saying I won't get around to it, but school just started and I don't really have so much extra time

MarcSabatella commented:
Thanks to @Jojo-Schmitz for tackling this in #2204; I've merged his branch into mine for #2202, and we'll see what happens.
